### PR TITLE
Polychromatic Riemann Sum

### PR DESCRIPTION
**Pull request by johanmazoyer**
+ This PR introduces the Riemann Sum to do polychromatic images from monochromatic images not necessarily evenly sampled. It introduce a new parameter in the `[modelconfig]` section, `mandatory_wls`. This optional keyword allows the introduction of specific wavelengths that need to appear to simulate the polychromatic images. I've made it a "hidden parameter" which means it's not in the parameter files but if you know what you're doing you can modify this. 
These new wavelengths must be in the range ]wavelength_0 - Delta_wav / 2 , wavelength_0 + Delta_wav / 2[. Default is an empty list ('mandatory_wls =  , '). In this case, the nb_wav simulation wavelengths are chosen to be well equally distributed in the Delta_wav. Following figure is in the docs to understand both case: 

image

We then do a Riemann sum to create the polychromatic image:
image

+ This new option allows more freedom in the choice of the estimation and correction wavelength. This PR introduces a new parameter in `[Estimationconfig]`: `estimation_wls`. This allows us to hand pick the correction and estimation wavelengths that we want, even if they are not well centered or evenly space in the bandwidth. I have checked that photon noise should be working. 

+ I modified THD2 class to automatically include `estimation_wls` in `mandatory_wls`.

+ Lot of changes in the docs to be a better documentation

+ Solve some errors in polychromatic correction. This is now quite well tested and I think this is working and ready for the testbed

+ up to v2.6

Docs for this branch can be found here : https://asterix-hci.readthedocs.io/en/polychrom_rieman_sum/

### CODE REVIEW

File: Asterix/tests/test_wfsc.py [hunk 1]
```diff
@@ -79,6 +80,9 @@ def test_1dm_correction():
     test_dir = get_data_dir(datadir="asterix_test_dir")
 
     best_contrast_1DM = quick_run_no_save(config, test_dir)
+
+    # be careful to take some margin in the final contrast because it might depends
```
Comment (ivalaginja) [inline Asterix/tests/test_wfsc.py:84]: typos in depend~s~

File: Asterix/tests/test_wfsc.py [hunk 2]
```diff
@@ -88,25 +92,56 @@ def test_2dm_correction():
...
-                                     'Nbiter_corr': [5],
-                                     'Nbmode_corr': [250]
-                                 })
+                                 NewLoopconfig={'Nbiter_corr': [7]})
 
     # we create a specific test_dir because we want the matrices to be redone everytime in testing
     test_dir = get_data_dir(datadir="asterix_test_dir")
     best_contrast_2DM = quick_run_no_save(config, test_dir)
 
+    # be careful to take some margin in the final contrast because it might depends
```
Comment (ivalaginja) [inline Asterix/tests/test_wfsc.py:109]: typos in depend~s~

File: Asterix/tests/test_wfsc.py [hunk 3]
```diff
@@ -88,25 +92,56 @@ def test_2dm_correction():
...
+    # on the phase aberrations in entrance that are randomly draw
```
Comment (ivalaginja) [inline Asterix/tests/test_wfsc.py:110]: typo in draw*n*

File: Asterix/tests/test_wfsc.py [hunk 4]
```diff
@@ -129,16 +164,15 @@ def test_2dm_correction_roman():
...
-                                     'Nbmode_corr': [250]
-                                 })
+                                 NewLoopconfig={'Nbiter_corr': [15]})
 
     # we create a specific test_dir because we want the matrices to be redone everytime in testing
     test_dir = get_data_dir(datadir="asterix_test_dir")
     best_contrast_2DM = quick_run_no_save(config, test_dir)
 
-    assert best_contrast_2DM < 1e-8, "Best contrast 2DM Roman HLC should be < 1e-8"
+    # be careful to take some margin in the final contrast because it might depends
```
Comment (ivalaginja) [inline Asterix/tests/test_wfsc.py:173]: typos in depend~s~

File: Asterix/tests/test_wfsc.py [hunk 5]
```diff
@@ -129,16 +164,15 @@ def test_2dm_correction_roman():
...
+    # on the phase aberrations in entrance that are randomly draw
```
Comment (ivalaginja) [inline Asterix/tests/test_wfsc.py:174]: typo in draw*n*

File: Asterix/tests/test_wfsc.py [hunk 6]
```diff
@@ -88,25 +92,56 @@ def test_2dm_correction():
...
+                                     'correction_algorithm': "sm",
+                                     'Nbmodes_OnTestbed': 600
+                                 },
+                                 NewLoopconfig={'Nbiter_corr': [7]})
+
+    # we create a specific test_dir because we want the matrices to be redone everytime in testing
+    test_dir = get_data_dir(datadir="asterix_test_dir")
+    best_contrast_2DM = quick_run_no_save(config, test_dir)
+
+    # be careful to take some margin in the final contrast because it might depends
```
Comment (ivalaginja) [inline Asterix/tests/test_wfsc.py:140]: typos in depend~s~

File: Asterix/tests/test_wfsc.py [hunk 7]
```diff
@@ -88,25 +92,56 @@ def test_2dm_correction():
...
+    # on the phase aberrations in entrance that are randomly draw
```
Comment (ivalaginja) [inline Asterix/tests/test_wfsc.py:141]: typo in draw*n*

File: docs/correction.rst [hunk 8]
```diff
@@ -148,13 +148,11 @@ Correction loop
...
 
-The ``correction_loop_1matrix()`` function is a loop running ``Nbiter_corr`` times that is basically doing:
+The ``correction_loop_1matrix()`` function is a loop running ``Nbiter_corr`` times. For each iteration, the following steps are done:
 
 * estimation
-
 * correction
-
-* application on DM and measure of DM
+* application on DM and measure of focal plane.
```
Comment (ivalaginja) [inline docs/correction.rst:155]: application to DM and measurement of focal plane

File: setup.py [hunk 9]
```diff
@@ -26,7 +26,7 @@
 
         # Specify the Python versions you support here. In particular, ensure
         # that you indicate whether you support Python 2, Python 3 or both.
-        'Programming Language :: Python :: 3.8, 3.9, 3.10',
+        'Programming Language :: Python :: 3.9, 3.10, 3.11',
```
Comment (ivalaginja) [inline setup.py:29]: Did you actually test all of these versions? Including a full conda install with these Python versions, from scratch?

**Review by ivalaginja — COMMENTED** [review]
Looks mostly fine to me but still trying to figure out how it works and whether it plays nice with the testbed controls.

File: Asterix/optics/optical_systems.py [hunk 10]
```diff
@@ -63,13 +63,64 @@ def __init__(self, modelconfig):
...
+            self.riemannn_intervals_lengths = np.diff(riemann_interval_edges)
+
         else:
+            # If we are in the monochromatic case, we define self.wav_vec[0] = wavelength_0
+            # and riemannn_intervals_lengths[0] = 1 so we can apply the same formula.
             self.wav_vec = np.array([self.wavelength_0])
+            self.riemannn_intervals_lengths = np.array([self.Delta_wav])
             self.nb_wav = 1
 
         self.string_os = '_dimPP' + str(int(self.dim_overpad_pupil)) + "_resFP" + str(round(
```
Comment (ivalaginja) [inline Asterix/optics/optical_systems.py:126]: Comment by @rgalicher on lines 126+127 :
The science_sampling is the one for the central wavelength but when matrices are saved for "estimation_wl", the sampling value should change in the filename. It seems the value changes for the calculation. This is just a naming thing.

**Comment by rgalicher** [comment]
Need to set automatically the wavelength for estimation when monochromatic Ignoring all useless keyword as estimation_wls nb_wav_estim polychromatic etc

**Comment by ivalaginja** [comment]
We were able to create matrices on this branch today that we could use to run on the testbed. Setup for THD2 in configfile:

Set central wavelength to shortest wavelength on testbed since this will create the smallest DH in terms of pixels, which is then used for all wavelengths:
`wavelength_0 = 637e-9`

Set the wavelength range large enough so that it covers your wavelength furthest away from central wavelength (in our case 637 --> 783 nm):
`Delta_wav = 300e-9`

Set the focal-plane sampling at the central wavelength:
`Science_sampling = 5.9`

Tell the code we want to run in multiple wavelengths:
`polychromatic = 'multiwl'`

Tell the code which wavelengths to include in the estimation since these are the wavelengths we need all matrices for:
`estimation_wls = 637e-9, 705e-9, 783.25e-9`

And in `estimator.py`, do the following change (see #163):
`if self.Estim_sampling < 3:` --> `if self.Estim_sampling <= 2:`

---------

This digs a decent DH in simulations and we manage to set up the testbed with this. For me this means this PR is good to go provided all comments get addressed and we tweak things for the testbed independently.

File: Asterix/Example_param_file.ini [hunk 11]
```diff
@@ -13,16 +13,16 @@ onbench=True
...
 
     # Number of monochromatic images in the spectral band.
     # Ignored if Delta_wav = 0
     # must be odd number
     nb_wav = 15
 
-    #Dimension of science image
-    dimScience   = 400  #detector science image size (in pixels)
+    # Dimension of science image
+    dimScience = 400  #detector science image size (in pixels)
```
[the remark this answers is not in the document]
Comment (johanmazoyer) [reply Asterix/Example_param_file.ini:25]: ok

File: Asterix/Example_param_file.ini [hunk 12]
```diff
@@ -225,23 +224,35 @@ onbench=True
...
-    # if polychromatic = 'multiwl'
-    # Number of monochromatic images for the estimation / correction. Must be an odd integer,
-    # divisor of nb_wav (parameter in [modelconfig])).
+    polychromatic = 'singlewl'
+
+    # We have 2 ways of defining the estimation / correction wavelengths. If polychromatic = 'broadband_pwprobes',
+    #  the central wavelength and bandwidth are always used. For other case, you can use 2 different methods :
+
+    # Method 1 (preferable for beginners) : automatic selection.
+    # If no estimation_wls are hand-picked "estimation_wls = "," the estimation / correction wavelengths are
```
[the remark this answers is not in the document]
Comment (johanmazoyer) [reply Asterix/Example_param_file.ini:243]: better ? 

    # If polychromatic = 'broadband_pwprobes', estimation is done using broadband images using
    # wavelength_0 and Delta_wav parameters and Correction is done at wavelength_0.
    # For other cases, you can use 2 different methods:

File: Asterix/Example_param_file.ini [hunk 13]
```diff
@@ -225,23 +224,35 @@ onbench=True
...
+    # If monochromatic images (Delta_wav = 0), all these options are equivalent
     # Not case sensitive
-    polychromatic = 'centralwl'
-
-    # if polychromatic = 'multiwl'
-    # Number of monochromatic images for the estimation / correction. Must be an odd integer,
-    # divisor of nb_wav (parameter in [modelconfig])).
+    polychromatic = 'singlewl'
+
+    # We have 2 ways of defining the estimation / correction wavelengths. If polychromatic = 'broadband_pwprobes',
```
[the remark this answers is not in the document]
Comment (johanmazoyer) [reply Asterix/Example_param_file.ini:239]: yes it is

File: Asterix/tests/param_file_tests.ini [hunk 14]
```diff
@@ -227,20 +227,32 @@ onbench=False
...
-    # Number of monochromatic images for the estimation / correction. Must be an odd integer,
-    # divisor of nb_wav (parameter in [modelconfig])).
+    polychromatic = 'singlewl'
+
+    # We have 2 ways of defining the estimation / correction wavelengths. If polychromatic = 'broadband_pwprobes',
+    #  the central wavelength and bandwidth are always used. For other case, you can use 2 different methods :
+
+    # Method 1 (preferable for beginners) : automatic selection.
+    # If no estimation_wls are hand-picked "estimation_wls = "," the estimation / correction wavelengths are
+    # automatically estimated.
```
[the remark this answers is not in the document]
Comment (johanmazoyer) [reply Asterix/tests/param_file_tests.ini:244]: ok

File: Asterix/tests/param_file_tests.ini [hunk 15]
```diff
@@ -227,20 +227,32 @@ onbench=False
...
+    # If monochromatic images (Delta_wav = 0), all these options are equivalent
     # Not case sensitive
-    polychromatic = 'centralwl'
-
-    # if polychromatic = 'multiwl'
-    # Number of monochromatic images for the estimation / correction. Must be an odd integer,
-    # divisor of nb_wav (parameter in [modelconfig])).
+    polychromatic = 'singlewl'
+
+    # We have 2 ways of defining the estimation / correction wavelengths. If polychromatic = 'broadband_pwprobes',
```
[the remark this answers is not in the document]
Comment (johanmazoyer) [reply Asterix/tests/param_file_tests.ini:239]: broadband_pwprobes still exists

**Comment by johanmazoyer** [comment]
> Need to set automatically the wavelength for estimation when monochromatic Ignoring all useless keyword as estimation_wls nb_wav_estim polychromatic etc

I think this is already the case ? Or am I mistaken

File: Asterix/optics/optical_systems.py (hunk 10 above)
Comment (johanmazoyer) [reply Asterix/optics/optical_systems.py:126]: ok done

File: setup.py (hunk 9 above)
Comment (johanmazoyer) [reply setup.py:29]: grblblb, yes I just did.

**Comment by johanmazoyer** [comment]
> We were able to create matrices on this branch today that we could use to run on the testbed. Setup for THD2 in configfile:
> 
> Set central wavelength to shortest wavelength on testbed since this will create the smallest DH in terms of pixels, which is then used for all wavelengths: `wavelength_0 = 637e-9`

Argh did not think of that. ok you tweaked it but mayeb we need to be more careful in our definition. For the moment, the DH is define unsing lambd0/D and the whole simulation part assume that lambda_0 is at the center of the correction BW 
I think it would be more in line with the code to actually modify the DH size than wavelength result, but equivalent for the testbed part

> Set the wavelength range large enough so that it covers your wavelength furthest away from central wavelength (in our case 637 --> 783 nm): `Delta_wav = 300e-9`
> 
> Set the focal-plane sampling at the central wavelength: `Science_sampling = 5.9`
> 
> Tell the code we want to run in multiple wavelengths: `polychromatic = 'multiwl'`
> 
> Tell the code which wavelengths to include in the estimation since these are the wavelengths we need all matrices for: `estimation_wls = 637e-9, 705e-9, 783.25e-9`
> 
> Make sure the number of wavelengths for the estimation is indeed what we give it in the line above (see also suggestion to make this adjustment automatically [here](https://github.com/johanmazoyer/Asterix/pull/159#issuecomment-1449910724)): `nb_wav_estim = 3`
> 

Done
> And in `estimator.py`, do the following change (see #163): `if self.Estim_sampling < 3:` --> `if self.Estim_sampling <= 2:`
> 

Done, I put 2.5. Please study at waht point we decrease performance.


> A truncation of the inverted control matrix at 700 modes seems to work ok but we didn't investigate much further.
> 
> This digs a decent DH in simulations and we manage to set up the testbed with this. For me this means this PR is good to go provided all comments get addressed and we tweak things for the testbed independently.

File: Asterix/Example_param_file.ini (hunk 12 above)
[the remark this answers is not in the document]
Comment (ivalaginja) [reply Asterix/Example_param_file.ini]: I think so

File: Asterix/wfsc/estimator.py [hunk 16]
```diff
@@ -64,50 +64,95 @@ def __init__(self, Estimationconfig, testbed: Testbed, matrix_dir='', save_for_b
...
+
+        # we measure the estimation sampling for the central wavelength wavelength_0.
         self.Estim_sampling = testbed.Science_sampling / Estimationconfig["Estim_bin_factor"]
 
-        if self.Estim_sampling < 3:
-            raise ValueError("Estim_sampling must be > 3, please decrease Estim_bin_factor parameter")
+        for wavei in self.wav_vec_estim:
+            if self.Estim_sampling / testbed.wavelength_0 * wavei < 2.5:
+                raise ValueError(f"Estimator sampling must be >= 2.5 at all estimation wavelengths, "
+                                 "please decrease [Estimationconfig]['parameter'] parameter")
```
Comment (ivalaginja) [inline Asterix/wfsc/estimator.py:145]: I think you forgot to fill in the two "parameter" here.

**Review by ivalaginja — COMMENTED** [review]
Looks good to me now. I'd like to take it for another spin on the testbed to make sure everything is fine before approving.

**Review by ivalaginja — DISMISSED** [review]
Just found another issue, when in broadband, the code does not save the results to `Mean_Contrast_DH.fits` properly.

**Comment by johanmazoyer** [comment]
> Just found another issue, when in broadband, the code does not save the results to `Mean_Contrast_DH.fits` properly.

ok, i'll do it in another PR because I kind of loose track of all the changes maid here

**Comment by ivalaginja** [comment]
@johanmazoyer can somebody confirm this works ok on hardware before I approve it? With working ok meaning, the output files are read ok by the Labview controls and it kinda digs a DH, even if not super well right now.

**Review by rgalicher — APPROVED** [review]

**Pull request merged**